daisy_workflows/linux_common/utils/guestfsprocess.py

```diff
@@ -0,0 +1,127 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Run a process in a guestfs.GuestFS instance.
+
+This library addresses a shortcoming of GuestFS.command and GuestFS.sh
+that cause data loss. When the command is successful, stderr is discarded.
+When the command fails, stdout is discarded.
+
+This library is motivated by subprocess.run, which returns an object
+containing stdout, stderr, and the return code of the process.
+"""
+
+import os
+import shlex
+import textwrap
+import typing
+
+
+def run(g: 'GuestFSInterface', command) -> 'CompletedProcess':
+  """Runs a process in a mounted GuestFS instance, ensuring that
+  standard output and standard error is always retained.
+
+  Args:
+    g: Mounted GuestFS instance.
+    command (str or List[str]): Script content that will be executed
+    by a bash interpeter on the guest.
+
+  Examples:
+    >>> run(g, 'date').stdout
+    Thu 05 Nov 2020 06:53:55 PM PST
+
+    >>> run(g, 'printf hi; exit 1')
+    {'stdout': 'hi', 'stderr': '', 'code': 1, 'cmd': 'printf hi; exit 1'}
+  """
+  tmp_dir = g.mkdtemp('gprocXXXXXX')
+  program_path = os.path.join(tmp_dir, 'program.sh')
+  stdout_path = os.path.join(tmp_dir, 'stdout.txt')
+  stderr_path = os.path.join(tmp_dir, 'stderr.txt')
+  return_code_path = os.path.join(tmp_dir, 'return_code.txt')
+
+  if not isinstance(command, str):
+    command = ' '.join(shlex.quote(s) for s in command)
+
+  program = _make_wrapping_program(command, stdout_path, stderr_path,
+                                   return_code_path)
+  g.write(program_path, program)
+  g.command(['/bin/bash', program_path])
+
+  return CompletedProcess(cmd=command,
+                          stdout=g.cat(stdout_path),
+                          stderr=g.cat(stderr_path),
+                          code=int(g.cat(return_code_path)))
+
+
+def _make_wrapping_program(
+    cmd: str, stdout_path: str, stderr_path: str,
+    return_code_path: str) -> str:
+  """Creates a shell script that captures the stdout, stderr, and return code
+  to files on the filesystem.
+
+  Args:
+    cmd: Command to execute.
+    stdout_path: Path to file for stdout.
+    stderr_path: Path to file for stderr.
+    return_code_path: Path to write return code of executing cmd.
+
+  Returns:
+    String containing the shell script.
+  """
+  return textwrap.dedent("""
+  touch {stdout_path}
+  touch {stderr_path}
+  touch {return_code_path}
+
+  bash -c {cmd} 1> {stdout_path} 2> {stderr_path}
```

Comment:
no quoting around cmd? have you validated this will work for all sorts of commands you want to call? what are the commands being run now?

Reply:
It's quoted by `shlex.quote` prior to being written to the template: https://docs.python.org/3/library/shlex.html#shlex.quote

See the  `test_escape_array_members`, for example.